### ISSUE TITLE
AsteroidsDeterministic-v0 (and v3) renders ship and score but not asteroids. NoFrameskip strobes.

### ISSUE THREAD

**Issue by mpgussert** [openai/gym#414]
Greetings all!  

Title says it all.  When using the Asteroids environment with Deterministic the score and ship are both rendered, but the actual asteroids are not.  Below is the minimum working example

```
import gym
  
  env = gym.make("AsteroidsDeterministic-v0")
  env.reset()
  for i in range(10000):
    env.render()
    env.step(env.action_space.sample())
```

also, If you use NoFrameskip, the asteroids strobe like mad.  

what's up with that?  I'm new to the gym so it's possible i just don't understand explicitly what the difference between Deterministic and NoFrameskip are.

Is there a simple fix?  I'm shooting for movement that looks like the standard asteroids env but without the random frame skipping.

Thanks!
Gus

**Comment by tlbtlbtlb** [comment]
The Atari games got a certain shimmering effect by drawing some objects only every 2nd frame. I assume it's drawing the asteroids that way. With Deterministic, it reports every 4th frame, so depending on initialization, you have a 50% chance of seeing asteroids. 

The difference is that NoFrameskip reports every frame, Determinstic reports every 4th frame, and the regular version reports every 2-5 frames (randomly chosen each step).

Obviously the current SpaceInvadersDeterminstic-v0 isn't useful. Possible solutions include:
 - Disallow the Deterministic version of this game
 - average N frames rather than taking the Nth. Which is roughly how it looks to humans.
 - report every 3rd frame instead of 4th (as with Space Invaders)

I think averaging would be best. Opinions?

**Comment by mpgussert** [comment]
averaging might work, but if there is a similar issue with space invaders and they just use the third frame, that might be the quickest fix.  where can I find the code that controls what frames are rendered when?  I checked the atari_env.py file which lead me to the rendering.py file in the gym, but there's nothing about frame skipping (that said, i am pretty oblivious.  I may have missed it).  where should I look next?  the documentation is unclear.

**Comment by mpgussert** [comment]
so after doing some testing.

in all cases it's not as simple as combining frames and rendering that way.  the actual environment class for asteroids probably needs to be changed.  The reason for this is that the only way to step through time in the render is to call the environment.step(...) function.  This function is responsible for returning the reward between the current frame (state) and the next state.  if you shoot at the right time, you can get the asteroid reward to increment DURING one of the off rendered frames.

This means if you want to catch the appropriate rewards you need to manually over write the states returned by the step function to be the frames you actually render to the screen.  You also need to overwrite the reward to be the sum of the two frames you combined (so it could be possible to earn more than 100 points for destroying a single asteroid on the screen if you destroyed one in the off rendered screen just before).  All of these changes are effectively changes to the environment itself (i am manually modifying the reward), which kinda goes against the philosophy of the gym in the first place...

TL;DR  this issue is not reasonably "work aroundable" and probably needs to be fixed in the gym source code.

**Comment by zplizzi** [comment]
This still seems to be an issue with the deterministic v4 Asteroids environment.

**Comment by christopherhesse** [comment]
Is this a problem when using the `NoFrameskip-v4` variants of the environments?  Flickering sprites appear to be a fundamental limitation of the Atari 2600, but using the no frameskip version should let you stack or max frames as desired.

**Comment by zplizzi** [comment]
I believe `NoFrameskip-v4` works as expected (flickering) - so agreed that you can handle it manually. It just catches people off guard and so perhaps would be nice to have a default version of the game that handled this for you (eg by only surfacing the frames with everything rendered properly). But maybe that's beyond scope.

**Comment by jkterry1** [comment]
Assignee: @JesseFarebro

**Comment by JesseFarebro** [comment]
As others have mentioned this is a limitation of the Atari 2600. Some objects aren't rendered on every frame. To combat this there have been many proposed solutions, e.g., colour averaging between frames, taking the max between frames, etc.

The most well agreed upon method used in practice is taking the max between frames. The `AtariPreprocessing` wrapper in Gym does this for you.
